docs: add `CONTRIBUTING.md` (#4134)
paritytech/rust-libp2p: CONTRIBUTING.md

Reviewer: Can you add a section like this please:

```md
### Self-service CI

We have a lot of automated CI checks for common errors.
Please treat our CI as a self-service platform and try to fix any issues before requesting a review.
```

# Contributing Guidelines

[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://ipn.io)
[![](https://img.shields.io/badge/project-libp2p-blue.svg?style=flat-square)](https://libp2p.io/)

Welcome to the rust-libp2p contribution guide! We appreciate your interest in improving our library. This guide will provide you with the necessary information to get started contributing to the project.

> ## Table of Contents

Reviewer: No ToC please.

Author: Done

    - [Reporting Issues](#reporting-issues)
    - [Submitting a feature request](#submitting-a-feature-request)
  - [Code Contribution Guidelines](#code-contribution-guidelines)
    - [Discuss big changes as Issues first](#discuss-big-changes-as-issues-first)
    - [Welcoming Pull Requests](#welcoming-pull-requests)
    - [Conventions](#conventions)
      - [Commit messages](#commit-messages)
      - [Commit message examples](#commit-message-examples)
      - [Mergify and the Send-it Label](#mergify-and-the-send-it-label)
      - [Writing Changelog Entries](#writing-changelog-entries)
      - [Bumping Versions](#bumping-versions)
    - [Code](#code)
    - [Commits](#commits)
    - [Tests](#tests)
    - [Documentation](#documentation)
    - [Pull Requests](#pull-requests)
      - [Code Review](#code-review)
      - [Rebasing](#rebasing)
      - [Merge Approval](#merge-approval)
      - [Reverting Changes](#reverting-changes)
      - [What if a commit that is supposed to be reverted contains changes that are also good?](#what-if-a-commit-that-is-supposed-to-be-reverted-contains-changes-that-are-also-good)
  - [Credits](#credits)

## Community Guidelines

> At rust-libp2p, we value a friendly and inclusive community. Harassment and abusive behavior are strictly prohibited. Violations of our community guidelines may result in immediate expulsion. Please be respectful and kind to fellow contributors.

Reviewer: Whilst good, I find this obvious so lets remove it.

Author: Done

## Looking for ways to contribute?

There are several ways you can contribute to rust-libp2p.

### Dive Right In

> If you're eager to start contributing immediately, check out our [help wanted](https://github.com/libp2p/rust-libp2p/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22) or [difficulty:easy](https://github.com/libp2p/rust-libp2p/issues?q=is%3Aissue+is%3Aopen+label%3Adifficulty%3Aeasy) issues on GitHub. These issues are suitable for newcomers and provide an excellent starting point.

Reviewer: I like this section.

### Reporting Issues

If you encounter bugs, mistakes, or inconsistencies in the rust-libp2p codebase or documentation, please let us know by creating an issue. No issue is too small, and your feedback is valuable in improving our project.

To facilitate the bug reporting process, we have provided a template that will assist you. If you believe you have encountered a bug, kindly complete the form to the best of your ability. Don't be concerned if you cannot provide every single detail, simply provide the information you can. Our contributors will ask for further clarification if anything is unclear.

For general questions and discussions, please visit our [discussion forum](https://github.com/libp2p/rust-libp2p/discussions). If you have already gone through the available documentation and still have unanswered questions or encounter difficulties, don't hesitate to seek assistance by initiating a discussion.

### Submitting a feature request

> If you wish to suggest a new feature, you can submit a feature request through the issue tracker. Simply fill out the provided form, ensuring to include a comprehensive explanation of the desired feature and any relevant context. Feel free to include examples of other tools that already have the feature you are requesting, if applicable.

Reviewer: I find these obvious too.

# Contributing Guidelines

[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://ipn.io)
[![](https://img.shields.io/badge/project-libp2p-blue.svg?style=flat-square)](https://libp2p.io/)

Welcome to the rust-libp2p contribution guide! We appreciate your interest in improving our library.

## Looking for ways to contribute?

There are several ways you can contribute to rust-libp2p:
- Start contributing immediately via the opened [help wanted](https://github.com/libp2p/rust-libp2p/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22) or [difficulty:easy](https://github.com/libp2p/rust-libp2p/issues?q=is%3Aissue+is%3Aopen+label%3Adifficulty%3Aeasy) issues on GitHub.
  These issues are suitable for newcomers and provide an excellent starting point.

> - Reporting issues, bugs, mistakes, or inconsistencies.

Reviewer: ```suggestion
- Reporting issues, bugs, mistakes, or inconsistencies.
  As many open source projects, we are short-staffed, we thus kindly ask you to be open to contribute a fix for discovered issues.
```

> - Submitting a feature request.
> - Opening a new pull request: primary means of making concrete changes to the code.

Reviewer: ```suggestion
```

Let's remove this.

> ### Code and Conventions

Reviewer: Let's remove this heading and flatten out everything below.

> #### Pull Requests

Reviewer: ```suggestion
#### We squash-merge pull Requests
```

#### Pull Requests

> To streamline the contribution process, we aim to simplify the PR template.
> Instead of including all the information in the template itself, we will provide here clear guidelines and references.
> 

Reviewer: ```suggestion
```

The PR title, which will become the commit message after the squashing process, should follow [conventional commit spec](https://www.conventionalcommits.org/en/v1.0.0/).

> #### Writing Changelog Entries

Reviewer: ```suggestion
#### Write changelog entries for user-facing changes
```

> The PR title, which will become the commit message after the squashing process, should follow conventional commit spec.

Reviewer: ```suggestion
The PR title, which will become the commit message after the squashing process, should follow [conventional commit spec](https://www.conventionalcommits.org/en/v1.0.0/).
```

#### Writing Changelog Entries

> When making significant changes, it is important to include corresponding entries in the changelog, providing a comprehensive summary for the users.

Reviewer: ```suggestion
When making user-facing changes, it is important to include corresponding entries in the changelog, providing a comprehensive summary for the users.
```

> For detailed instructions on how to write changelog entries, please refer to the documentation in `docs/release.md`.

Reviewer: Let's make `docs/release.md` a link please.

> #### Bumping Versions
> 
> To maintain consistency and clarity in our releases, it's important to correctly update the version of the relevant crate when making changes that require changelog entries.
> Detailed instructions on how to update versions can be found in the documentation available at [`docs/release.md`](https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md).

Reviewer: ```suggestion
```

This is documented in `release.md`.

> #### Mergify and the Send-it Label

Reviewer: ```suggestion
#### Merging of PRs is automated
```

To maintain consistency and clarity in our releases, it's important to correctly update the version of the relevant crate when making changes that require changelog entries.
Detailed instructions on how to update versions can be found in the documentation available at `docs/release.md`.

> Accurate versioning is crucial to ensure proper tracking of changes.

Reviewer: I think this is actually explained well in `docs/release.md` so we can omit this here :)

#### Mergify and the Send-it Label

To streamline our workflow, we utilize Mergify and the "send-it" label.
Mergify automates merging actions and helps us manage pull requests more efficiently.
The "send-it" label indicates that a pull request is ready to be merged.

> Please refrain from making further commits after the "send-it" label has been applied.

Reviewer: ```suggestion
Please refrain from making further commits after the "send-it" label has been applied otherwise your PR will be dequeued from merging automatically.
```

> ### Self-service CI

Reviewer: ```suggestion
### Treat CI as a self-service platform
```

# Contributing Guidelines

[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://ipn.io)
[![](https://img.shields.io/badge/project-libp2p-blue.svg?style=flat-square)](https://libp2p.io/)

Welcome to the rust-libp2p contribution guide! We appreciate your interest in improving our library.

## Table of Contents

- [Contributing Guidelines](#contributing-guidelines)
  - [Table of Contents](#table-of-contents)
  - [Community Guidelines](#community-guidelines)
  - [Looking for ways to contribute?](#looking-for-ways-to-contribute)
  - [Code Contribution Guidelines](#code-contribution-guidelines)
    - [Discuss big changes as Issues first](#discuss-big-changes-as-issues-first)
    - [Welcoming Pull Requests](#welcoming-pull-requests)
    - [Code and Conventions](#code-and-conventions)
      - [Commits](#commits)
      - [Pull Requests](#pull-requests)
      - [Tests](#tests)
      - [Documentation](#documentation)
      - [Writing Changelog Entries](#writing-changelog-entries)
      - [Bumping Versions](#bumping-versions)
      - [Mergify and the Send-it Label](#mergify-and-the-send-it-label)
  - [Credits](#credits)

## Community Guidelines

> At rust-libp2p, we value a friendly and inclusive community. Harassment and abusive behavior are strictly prohibited. Please be respectful and kind to fellow contributors.

Reviewer: Despite being good, I'd like to think that this is obvious so there is no need to mention it :)

## Looking for ways to contribute?

There are several ways you can contribute to rust-libp2p:

> - Start contributing immediately via the opened [issues](https://github.com/libp2p/rust-libp2p/issues).

Reviewer: I liked that you mentioned our labels before, I think that is useful.

- Reporting issues, bugs, mistakes, or inconsistencies.
- Submitting a feature request.
- Opening a new pull request: primary means of making concrete changes to the code.

## Code Contribution Guidelines

### Discuss big changes as Issues first

> Before starting to code significant improvements, it's important to document them as GitHub issues in order to get feedback.

Reviewer: I think we can remove that, I am not fussed and it is somewhat obvious to me.

### Welcoming Pull Requests

> Pull requests are the primary means of making concrete changes to the code, documentation, and dependencies of rust-libp2p. Even small pull requests, such as fixing a typo, are highly appreciated.

Reviewer: Same here, I don't think that is worth mentioning.

### Code and Conventions

#### Commits

It is highly recommended to organize your changes into logically grouped commits. Keeping each commit focused on a specific aspect of the changes makes it easier for reviewers to understand and evaluate the modifications.

> Commit messages must start with a short subject line, followed by an optional, more detailed explanatory text which is separated from the summary by an empty line.

Reviewer: We don't care much about the commits within a PR actually because we squash merge.

#### Pull Requests

> To streamline the contribution process, we aim to simplify the PR template. Instead of including all the information in the template itself, we will provide clear guidelines and references to these new contributor guidelines. This approach helps maintain a concise and focused PR template while providing contributors with comprehensive instructions.

Reviewer: Not sure what you are trying to say with this sentence.

> When submitting pull requests, we strongly encourage squash merging. This means that we discourage force pushes, as it makes the diff between pushes easier for us to review. Squash merging allows us to maintain a clean and organized commit history.

Reviewer: We don't just encourage squash-merging, we always squash-merge :)

We don't want force pushes because it messes with GitHub's review system.

> The PR title will become the commit message after the squashing process. By adhering to this convention, it becomes easier to track and understand the purpose of each commit.

Reviewer: I think it is enough here to say that the title should follow conventional commit spec. Please link to it.

#### Tests

> When making changes to the codebase, whether it involves introducing new functionality or fixing existing functionality, it is crucial to include one or more tests in your pull request. Take a look at existing tests for inspiration.

Reviewer: IMO this is kind of obvious too. Let's not clutter this document with it.

#### Documentation

> Update documentation when creating or modifying features. Test your documentation changes for clarity, concision, and correctness, as well as a clean documentation build.

Reviewer: Same here, I think it is okay to request this on-demand.

#### Writing Changelog Entries

> When making significant changes, it is important to include corresponding entries in the changelog. The changelog provides a summary of notable updates and serves as a reference for users and contributors. For detailed instructions on how to write changelog entries, please refer to the documentation in `docs/release.md`.

Reviewer: It is not the significance that is important but whether they are user-facing.

#### Bumping Versions

> When introducing changes that warrant a version update, it is essential to follow the correct versioning process. Detailed guidelines on how to bump versions can be found in the documentation located at `docs/release.md`. It is crucial to ensure accurate versioning to maintain clarity and consistency in our releases.

Reviewer: The rule of thumb is: changelog entry means version bump.

#### Mergify and the Send-it Label

> To streamline our workflow, we utilize Mergify and the "send-it" label. Mergify automates merging actions and helps us manage pull requests more efficiently. The "send-it" label indicates that a pull request is ready to be merged.

Reviewer: Mention here that people should _not_ push commits after `send-it` has been applied.

## Credits

> This document is based on [Contributing to IPFS](https://github.com/ipfs/community/blob/master/CONTRIBUTING.md).

Reviewer: Do we still need to keep this after the above modifications?

> This document is based on [Contributing to IPFS](https://github.com/ipfs/community/blob/master/CONTRIBUTING.md).

Reviewer: Do we still need to keep this after the above modifications?
